README.md

```diff
@@ -1 +1,264 @@
+                              | ...                        |
+                              | ...                        |
+                              \_...                        |
+                                e0000: mov $2, $eax        |
+                                ...                        |
+                                e0100: call implementation /
+                                ...                       /
+                                e0200: jmp db2aa ________/
+```
+*Hotpatching using a trampoline jump:*
+Replace a syscall instruction with a short jmp instruction,
+the destination of which is a regular jmp instruction.
+The reason to use this, is that a short jmp instruction
+consumes only two bytes, thus fits in the place of a syscall
+instruction. Sometimes the instructions directly preceding
+or following the syscall instruction can not be overwritten,
+leaving only the two bytes of the syscall instruction
+for patching.
+The hotpatching library looks for place for the trampoline jump
+in the padding found to the end of each routine. Since the start
+of all routines is aligned to 16 bytes, often there is a padding
+space between the end of a symbol, and the start of the next symbol.
+In the example below, this padding is filled with 7 byte long
+nop instruction (so the next symbol can start at the address 3f410).
+```
+Before:                         After:
+
+3f3fe: mov %rdi, %rbx           3f3fe: mov %rdi, %rbx
+3f401: syscall                /-3f401: jmp 3f430
+3f403: jmp 3f415              | 3f403: jmp 3f415 ----------\
+3f407: retq                   | 3f407: retq                |
+                              \                            |
+3f408: nopl 0x0(%rax,%rax,1)  /-3f408: jmp e1000           |
+                              | ...                        |
+                              | ...                        |
+                              \_...                        |
+                                e1000: nop                 |
+                                ...                        |
+                                e1100: call implementation /
+                                ...                       /
+                                e1200: jmp 3f403 ________/
+
+```
+
+# Limitations: #
+* Only Linux is supported
+* Only x86\_64 is supported
+* Only tested with glibc, although perhaps it works
+with some other libc implementations as well
+* There are known issues with the following syscalls:
+  * clone
+  * rt_sigreturn
+
+# Debugging: #
+Besides logging, the most important factor during debugging is to make
+sure the syscalls in the debugger are not intercepted. To achieve this, use
+the INTERCEPT_HOOK_CMDLINE_FILTER variable described above.
+
+```
+INTERCEPT_HOOK_CMDLINE_FILTER=ls \
+	LD_PRELOAD=libsyscall_intercept.so \
+	gdb ls
+```
+
+With this filtering, the intercepting library is not activated in the gdb
+process itself.
```

Comment:
The README should include requirements for any additional pkgs that need to be installed. Such as capstone. Whatever is required for the user to set up their system to run this.